core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,235 @@
+package com.mesosphere.usi.core
+
+import java.time.Instant
+
+import com.mesosphere.usi.core.models._
+
+case class FrameWithEffects(frame: Frame, effects: FrameEffects, dirtyPodIds: Set[PodId]) {
+  def applyEffects(newEffects: FrameEffects): FrameWithEffects = {
+    // TODO - we need to handle status snapshots and create a mechanism to signal that all cache should be recomputed
+    val newDirty = dirtyPodIds ++ newEffects.reverseStateEvents.iterator.collect {
+      case podEvent: PodStateEvent => podEvent.id
+    }
+    copy(frame = frame.applyStateEffects(newEffects), dirtyPodIds = newDirty, effects = effects ++ newEffects)
+  }
+
+  def applySpecEvent(specEvent: SpecEvent): FrameWithEffects = {
+    // TODO - assert valid transition
+    specEvent match {
+      case SpecsSnapshot(podSpecSnapshot, reservationSpecSnapshot) =>
+        if (reservationSpecSnapshot.nonEmpty) {
+          // This should make the framework crash
+          throw new NotImplementedError("ReservationSpec support not yet implemented")
+        }
+        val newPodsSpecs: Map[PodId, PodSpec] = podSpecSnapshot.map { pod =>
+          pod.id -> pod
+        }(collection.breakOut)
+
+        val changedPodIds = frame.podSpecs.keySet ++ newPodsSpecs.keySet
+
+        copy(frame = frame.copy(podSpecs = newPodsSpecs), dirtyPodIds = dirtyPodIds ++ changedPodIds)
+
+      case PodSpecUpdated(id, newState) =>
+        val newPodSpecs = newState match {
+          case Some(podSpec) =>
+            frame.podSpecs.updated(id, podSpec)
+          case None =>
+            frame.podSpecs - id
+        }
+
+        copy(frame = frame.copy(podSpecs = newPodSpecs), dirtyPodIds = dirtyPodIds ++ Set(id))
+
+      case ReservationSpecUpdated(id, _) =>
+        throw new NotImplementedError("ReservationSpec support not yet implemented")
+    }
+  }
+
+  def process(fn: FrameWithEffects => FrameEffects): FrameWithEffects = applyEffects(fn(this))
+}
+
+/**
+  * Container class responsible for keeping track of the frame and cache.
+  *
+  * As a general rule, business logic does not manipulate the USI state directly, but rather does so by returning
+  * effects.
+  */
+private[core] class SchedulerLogic {
+  private var frame: Frame = Frame.empty
+  private var cachedPendingLaunch = CachedPendingLaunch(Set.empty)
+
+  def processSpecEvent(msg: SpecEvent): FrameEffects = {
+    handleFrame { frameWithEffects =>
+      frameWithEffects
+        .applySpecEvent(msg)
+        .process { f =>
+          SchedulerLogic.computeNextStateForPods(f.frame)(f.dirtyPodIds)
+        }
+    }
```

Comment:
Here's my suggestion to flatten it

```scala
val (frameWithSpecEventsApplied, dirityPods) = applySpecEvent(currentFrame, msg)
val FrameIntents(stateUpdates, calls) = SchedulerLogic.computeNextStateForPods(frameWithSpecEventsApplied)(dirtyPodIds)

val finalFrame = applyIntents(stateUpdates)
currentFrame = finalFrame
FrameIntents(stateUpdates, calls)
```
The names are a little verbose.

Reply:
I really prefer the method chaining approach; it gets rid of all of these assignments and precludes us from accidentally returning the wrong version of the frame. Can you help me understand why you think this is better?